backend/compact-connect/stacks/persistent_stack/ssn_table.py

```diff
@@ -0,0 +1,154 @@
+    ServicePrincipal,
+    StarPrincipal,
+)
+from aws_cdk.aws_kms import Key
+from cdk_nag import NagSuppressions
+from common_constructs.stack import Stack
+from constructs import Construct
+
+
+class SSNTable(Table):
+    """DynamoDB table to house provider Social Security Numbers"""
+
+    def __init__(
+        self,
+        scope: Construct,
+        construct_id: str,
+        *,
+        removal_policy: RemovalPolicy,
+        **kwargs,
+    ):
+        self.key = Key(
+            scope,
+            'SSNKey',
+            enable_key_rotation=True,
+            alias='ssn-key',
+            removal_policy=removal_policy,
+        )
+
+        super().__init__(
+            scope,
+            construct_id,
+            # Forcing this resource name to comply with a naming-pattern-based CloudTrail log, to be
+            # implemented in issue https://github.com/csg-org/CompactConnect/issues/397
+            table_name='ssn-table-DataEventsLog',
+            encryption=TableEncryption.CUSTOMER_MANAGED,
+            encryption_key=self.key,
+            billing_mode=BillingMode.PAY_PER_REQUEST,
+            removal_policy=removal_policy,
+            point_in_time_recovery=True,
+            partition_key=Attribute(name='pk', type=AttributeType.STRING),
+            sort_key=Attribute(name='sk', type=AttributeType.STRING),
+            resource_policy=PolicyDocument(
+                statements=[
+                    PolicyStatement(
+                        # No actions that involve reading/writing more than one record at a time. In the event of a
+                        # compromise, this slows down a potential data extraction, since each record would need to be
+                        # pulled, one at a time
+                        effect=Effect.DENY,
+                        actions=[
+                            'dynamodb:BatchGetItem',
+                            'dynamodb:BatchWriteItem',
+                            'dynamodb:PartiQL*',
+                            'dynamodb:Query',
+                            'dynamodb:Scan',
+                        ],
+                        principals=[StarPrincipal()],
+                        resources=['*'],
+                        conditions={
+                            'StringNotEquals': {
+                                # We will allow DynamoDB itself, so it can do internal operations like backups
+                                'aws:PrincipalServiceName': 'dynamodb.amazonaws.com',
+                            }
+                        },
+                    )
+                ]
+            ),
+            **kwargs,
+        )
+        self.inverted_index_name = 'inverted'
+
+        self.add_global_secondary_index(
+            index_name=self.inverted_index_name,
+            partition_key=Attribute(name='sk', type=AttributeType.STRING),
+            sort_key=Attribute(name='pk', type=AttributeType.STRING),
+            projection_type=ProjectionType.ALL,
+        )
```

Comment:
It would be useful to document here the purpose of this GSI and the access patterns it supports.